### ISSUE TITLE
Regression in 3.1.3: false alarm for "Parameter must be non-null but is marked as nullable"

### ISSUE THREAD

**Issue by PhilippWendler** [spotbugs/spotbugs#616]
For this code:
```
import com.google.common.base.Preconditions;
import javax.annotation.Nullable;

public enum TestClass {
  TEST("one", "two", null);

  public final String one;
  public final String two;
  @Nullable public final String three;

  private TestClass(String one, String two, @Nullable Object three) {
    this.one = Preconditions.checkNotNull(one);
    this.two = Preconditions.checkNotNull(two);
    this.three = three == null ? null : three.toString();
  }
}
```

SpotBugs 3.1.3 outputs this warning:

```
Parameter must be non-null but is marked as nullable

In class TestClass
In method new TestClass(String, int, String, String, Object)
Parameter one
At TestClass.java:[lines 11-15]
one must be non-null but is marked as nullable
```

SpotBugs 3.1.2 does not show this warning. The code is compiled for Java 8.

This has to be a bug, because if I remove either parameter `two` or `three`, the warning about parameter `one` goes away.

**Comment by KengoTODA** [comment]
I cannot reproduce this reported bug by unit test, see 605fc5a for detail.

**Comment by PhilippWendler** [comment]
Thanks for investigating this!

In general, we have a `package-info.java` with annotations, but not in the default package where I put the test class.

I have uploaded my `.class` file here: [Issue616.zip](https://github.com/spotbugs/spotbugs/files/1965950/Issue616.zip)
Can you reproduce it with this?

Is the class path relevant for this kind of problem?
I have (among others) Guava 23.6-jre, spotbugs-annotations 3.1.1 and jsr305 3.0.2 from com.google.code.findbugs on the auxiliary classpath.

**Comment by FH3095** [comment]
Hello,

I today encountered the same problem as PhilippWendler . To me it seems to be relevant to mark the enum members NonNull to trigger the false positive. So when I modify your Issue616.java to the following
```
package com.example;

import java.util.Objects;

import edu.umd.cs.findbugs.annotations.NonNull;
import edu.umd.cs.findbugs.annotations.Nullable;

public enum Issue616 {
	TEST("one", "two", null);

	public final @NonNull String one;
	public final @NonNull String two;
	@Nullable
	public final String three;

	private Issue616(@NonNull String one, @NonNull String two, @Nullable Object three) {
		this.one = Objects.requireNonNull(one);
		this.two = Objects.requireNonNull(two);
		this.three = three == null ? null : three.toString();
	}
}

```
I get
```
Bug: one must be non-null but is marked as nullable

This parameter is always used in a way that requires it to be non-null, but the parameter is explicitly annotated as being Nullable. Either the use of the parameter or the annotation is wrong. 

Rank: Troubling (14), confidence: High
Pattern: NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE 
Type: NP, Category: STYLE (Dodgy code)
```
I'm not sure that this is the same problem but it is at least very similar.
@PhilippWendler Do you have some setting to make `@NonNull` the default?

Greetings

**Comment by KengoTODA** [comment]
I tried but this issue is still not reproducible. See d2eedeb43 for detail.

**Comment by PhilippWendler** [comment]
@FH3095 I do not know how I could make `@NonNull` the default for SpotBugs (we use package annotations, but not in the package I have tested this in).

@KengoTODA Can you reproduce the problem with my `.class` file?

**Comment by KengoTODA** [comment]
I tried but it's not reproducible.

You can fork my `issue-616` branch in this repo, and edit `spotbugsTestCases/src/java/ghIssues/Issue616.java` to have a try.

**Comment by PhilippWendler** [comment]
@KengoTODA I know you tried with the Java code you have linked, but what about my `.class` file? Can you reproduce it with that?

I think we should first find out if there is something in SpotBugs that lets it return different results for the same `.class` file on our machines, or if there are differences in the `.class` files that produce the difference in the SpotBugs results. If we find this out, we know where to look further.

**Comment by KengoTODA** [comment]
> but what about my .class file? Can you reproduce it with that?
As I said, I tried but it's not reproducible.

**Comment by FH3095** [comment]
Sadly I'm out of ideas.
I thought it probably is related to the jdk version so i switched to my linux machine with
```
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-1~deb9u1-b11)
OpenJDK 64-Bit Server VM (build 25.171-b11, mixed mode)
```
and used the following commands:
```
wget http://repo.maven.apache.org/maven2/com/github/spotbugs/spotbugs/3.1.3/spotbugs-3.1.3.tgz
tar -x -f spotbugs-3.1.3.tgz
javac -classpath spotbugs-3.1.3/lib/spotbugs-annotations.jar:spotbugs-3.1.3/lib/jsr305-3.0.2.jar Issue616.java
java -jar spotbugs-3.1.3/lib/spotbugs.jar -output a.xml -xml -auxclasspath spotbugs-3.1.3/lib/spotbugs-annotations.jar:spotbugs-3.1.3/lib/jsr305-3.0.2.jar Issue616.class
```
and got the expected "Warnings generated: 1" message.
I have no clue left why on CI it doesn't work.

[Files.zip](https://github.com/spotbugs/spotbugs/files/2002273/Files.zip)

**Comment by KengoTODA** [comment]
@FH3095 could you try with SpotBugs 3.1.2? I want to confirm that this problem is regression of 3.1.3 or not.

**Comment by FH3095** [comment]
@KengoTODA Seems to be a regression of 3.1.3. I used my linux machine again with the same commands as above just changed everything to 3.1.2  and got no warning.

[Files2.zip](https://github.com/spotbugs/spotbugs/files/2005981/Files2.zip)

**Comment by KengoTODA** [comment]
note: differences between 3.1.2 and 3.1.3 https://github.com/spotbugs/spotbugs/compare/3.1.2...3.1.3

**Comment by artylan** [comment]
I have false positives of this type with ctors of inner classes, and to me, it seems that it makes an off-by-one error due to the synthetic "this$0" argument.

**Comment by KengoTODA** [comment]
In my guess, #607 is not related with this problem, then ASM 6.1.1 or BCEL 6.2 could be the cause.

* ASM
    * [unreleased changes](https://gitlab.ow2.org/asm/asm/compare/ASM_6_1_1...master)
    * [issue tracker](https://gitlab.ow2.org/asm/asm/issues?scope=all&utf8=%E2%9C%93&state=all)
* BCEL
    * [unreleased changes](https://github.com/apache/commons-bcel/compare/master...BCEL_6_2)
    * [issue tracker](https://issues.apache.org/jira/browse/BCEL-299?jql=project%20%3D%20BCEL%20AND%20affectedVersion%20in%20(EMPTY%2C%206.2)%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC)

Anyway firstly we should have automated way to reproduce issue, then we can debug & test.

**Comment by zolyfarkas** [comment]
I am experiencing something potentially related with code (spotbugs 3.1.3.1):

 public contructor(@Nullable final String schema) {
    final String useSchema = Strings.isNullOrEmpty(schema) ? null : schema.toUpperCase(Locale.US);
....

and getting:

schema must be non-null but is marked as nullable [com.pimco.popcorn.core.jdbc.MetaData] At MetaData.java:[lines 133-208] NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE

**Comment by KengoTODA** [comment]
@zolyfarkas can you reproduce it with version 3.1.5?

**Comment by KengoTODA** [comment]
I'll close this stale issue, please feel free to reopen.

**Comment by PhilippWendler** [comment]
I can confirm the original issue (with the code from my first post) on SpotBugs 3.1.5.

Strangely, GitHub does not offer me the "Reopen button", please reopen @KengoTODA .

**Comment by olegshtch** [comment]
Same happened on SpotBugs 3.1.12

**Comment by PhilippWendler** [comment]
We switched to Java 11 and I can no longer reproduce the problem there. So I am closing this issue.

**Comment by jpschewe** [comment]
I'm seeing the issue show up under Java 11 when using the checker framework. Even adding the checker NonNull annotation to the parameter doesn't help, spotbugs still insists that the parameter is explicitly marked nullable.
I'm using spotbugs 4.1.3.

**Comment by bahtiyarkaba** [comment]
> I'm seeing the issue show up under Java 11 when using the checker framework. Even adding the checker NonNull annotation to the parameter doesn't help, spotbugs still insists that the parameter is explicitly marked nullable.
I'm using spotbugs 4.1.3.

I'm having the same issue.
